semgrep/requirements.txt

```diff
@@ -1,5 +1,6 @@
 colorama==0.4.3
 pytest==5.3.5
+pytest-snapshot==0.4.0
```

Comment:
Is there any way we can achieve the same level of testing without snapshotting? If I'm understanding them correctly, they seem very difficult to reason about. I can never tell why a huge JSON blob is no longer the same as a previous huge JSON blob. IMO we should testing small amounts of functionality and return values instead of dumping everything to a file and essentially performing a `diff`.

Reply:
This seems like the age-old debate about end-to-end vs. unit tests. IMO, in general you need both: end-to-end tests (using snapshots here) guarantee behavior, while unit tests allow you to debug regressions.

But end-to-end tests would usually be primary, as without them, you can't guarantee your system works. So we want those most urgently.

Reply:
cf. https://blog.colinbreck.com/should-we-write-a-unit-test-or-an-end-to-end-test/

Reply:
Thanks! I had similar thoughts and had just documented that philosophy here: https://github.com/returntocorp/semgrep/tree/bence/pytest/semgrep/tests

Reply:
FWIW, pretty-printing the JSON blobs might make debugging e2e failures much easier. Food for thought.

Reply:
Yeah. I was considering that too. Added this change!

Reply:
When I think about testing I tend to use unit tests to test a single component, integration tests to test the interaction of two (or a few) components, and e2e tests to test many components or the system as a whole. Since semgrep is really only one component, or two if you count the core separately, I would lean toward unit tests.

> ... end-to-end tests (using snapshots here) guarantee behavior, while unit tests allow you to debug regressions ... But end-to-end tests would usually be primary, as without them, you can't guarantee your system works.

I'm not sure I agree here. In my opinion unit tests can also guarantee behavior, and e2e tests can be used to debug regressions. I've seen highly stable systems that use nothing but unit tests, and very fragile systems that use all e2e tests, or a mix of all three.

Regardless, I won't harp on the issue too much since it seems the decision's been made to go with this approach :+1: